cpp/src/randomforest/randomforest.cu

```diff
@@ -341,6 +347,102 @@ std::vector<unsigned char> save_model(ModelHandle model) {
+  tl::Model& concat_model = *(tl::Model*)concat_tree_handle;
+  for (int tl_num = 0; tl_num < treelite_handles.size(); tl_num++) {
+    const tl::Model& model = *(tl::Model*)(treelite_handles[tl_num]);
+    bool check_num_features = concat_model.num_feature != model.num_feature;
+    bool check_num_output_group =
+      concat_model.num_output_group != model.num_output_group;
+    bool check_rf_flag =
+      concat_model.random_forest_flag != model.random_forest_flag;
+    if (check_num_features || check_num_output_group || check_rf_flag) {
+      throw(
+        "Error! the trees saved in concatenated forest and the different "
+        "from the ones present in the individual forests");
+    }
+
+    for (int indiv_trees = 0; indiv_trees < model.trees.size(); indiv_trees++) {
+      const tl::Tree concat_tree =
+        concat_model.trees[concat_mod_tree_num + indiv_trees];
+      const tl::Tree single_worker_tree = model.trees[indiv_trees];
+      if (concat_tree.num_nodes != single_worker_tree.num_nodes) {
+        throw(
+          "Error! the trees saved in concatenated forest and the different "
+          "from the ones present in the individual forests");
+      }
+
+      for (int each_node = 0; each_node < concat_tree.num_nodes; each_node++) {
+        const tl::Tree::Node& node_concat = concat_tree[each_node];
+        const tl::Tree::Node& node_single = single_worker_tree[each_node];
+        bool check_is_root = node_concat.is_root() != node_single.is_root();
+        bool check_parent = node_concat.parent() != node_single.parent();
+        bool check_is_leaf = node_concat.is_leaf() != node_single.is_leaf();
+        bool check_leaf = node_concat.leaf_value() != node_single.leaf_value();
+        bool check_cright = node_concat.cright() != node_single.cright();
+        bool check_cleft = node_concat.cleft() != node_single.cleft();
+        bool check_split =
+          node_concat.split_index() != node_single.split_index();
+
+        if (check_is_root || check_parent || check_is_leaf || check_leaf ||
+            check_split || check_cleft || check_cright) {
+          throw(
+            "Error! the trees saved in concatenated forest and the different "
+            "from the ones present in the individual forests");
+        }
+      }
+    }
+    concat_mod_tree_num = concat_mod_tree_num + model.trees.size();
+  }
+}
+
+std::vector<unsigned char> concatenate_trees(
+  const cumlHandle& handle, std::vector<ModelHandle*> treelite_handles) {
+  //tl::Model check_mod;
+  tl::Model& first_model = *(tl::Model*)treelite_handles[0];
+  tl::Model concat_model;
+  for (int tl_num = 0; tl_num < treelite_handles.size(); tl_num++) {
+    tl::Model& model = *(tl::Model*)treelite_handles[tl_num];
+    int limit = model.trees.size();
+    //auto copied_tree = model.trees[0];
+    for (int i = 0; i < limit; i++) {
+      (concat_model.trees).push_back((model.trees[i]));
+    }
+  }
+
+  ModelHandle concat_mod_handle = &(concat_model);
```

Comment:
It would be much cleaner to separate the tree concatenation from the conversion to protobuf bytes. If we think about "what's the one thing this function does," right now we'd say "it concatenates models and converts them to protobuf," which means it actually does two things.
If the caller wants protobuf bytes, they can always call `save_to_protobuf` themselves